make.py

```diff
@@ -132,24 +140,52 @@ def make_bilingual_book(self):
         )
         print("TODO need process bar here: " + str(all_p_length))
         index = 0
-        for i in self.origin_book.get_items():
-            if i.get_type() == 9:
-                soup = bs(i.content, "html.parser")
-                p_list = soup.findAll("p")
-                is_test_done = IS_TEST and index > 20
-                for p in p_list:
-                    if not is_test_done:
-                        if p.text and not p.text.isdigit():
-                            new_p = copy(p)
-                            # TODO banch of p to translate then combine
-                            # PR welcome here
+        p_t_len = len(self.p_t)
+        try:
+            for i in self.origin_book.get_items():
+                if i.get_type() == 9:
+                    soup = bs(i.content, "html.parser")
+                    p_list = soup.findAll("p")
+                    is_test_done = IS_TEST and index > 20
+                    for p in p_list:
+                        if is_test_done or not p.text or p.text.isdigit():
+                            continue
+                        new_p = copy(p)
+                        if self.resume and index < p_t_len:
+                            new_p.string = self.p_t[index]
+                        else:
                             new_p.string = self.translate_model.translate(p.text)
-                            p.insert_after(new_p)
-                            index += 1
-                i.content = soup.prettify().encode()
-            new_book.add_item(i)
-        name = self.epub_name.split(".")[0]
-        epub.write_epub(f"{name}_bilingual.epub", new_book, {})
+                            self.p_t.append(new_p.text)
+                        p.insert_after(new_p)
+                        index += 1
+                    i.content = soup.prettify().encode()
+                new_book.add_item(i)
+            name = self.epub_name.split(".")[0]
+            epub.write_epub(f"{name}_bilingual.epub", new_book, {})
+        except KeyboardInterrupt as e:
+            print(e)
+            print("you can resume it next time")
+            self.save_progress()
+            exit(0)
+        except Exception as e:
+            print(e)
+            print("you can resume it next time")
+            self.save_progress()
+            exit(0)
```

Comment:
can this Exception into one?
https://docs.python.org/3/tutorial/errors.html#handling-exceptions